### ISSUE TITLE
Crash under full load: NIO-ELT-#1 (5): EXC_BAD_ACCESS (code=EXC_I386_GPFLT)

### ISSUE THREAD

**Issue by andreigherghe** [vapor/fluent#487]
On any save operation, using FluentMySQL or FluentPostgreSQL, calling the route in an infinite loop several times in background processes results in `NIO-ELT-#1 (5): EXC_BAD_ACCESS (code=EXC_I386_GPFLT)`.
<img width="962" alt="screen shot 2018-05-19 at 00 45 13" src="https://user-images.githubusercontent.com/18578250/40261569-0f022174-5b0a-11e8-94a0-3e7c8f4032d7.png">
<img width="854" alt="screen shot 2018-05-19 at 00 01 44" src="https://user-images.githubusercontent.com/18578250/40261570-0f2627fe-5b0a-11e8-9331-b8f4a60af9ec.png">

Sample loop script:
```
while [ 1 ]
do
curl --request POST \
  --url http://localhost:8080/polls \
  --header 'content-type: application/json' \
  --data '{
                "question": "Okay?",
                "startDate": 123,
                "endDate": 456
}'
done
```
After enabling Thread Sanitizer, I got this additional piece of info: 
```
WARNING: ThreadSanitizer: data race (pid=98333)
  Write of size 8 at 0x000101bcc320 by thread T5:
    #0 RFC1123DateCache.currentTimestamp() RFC1123.swift:64 (HTTP:x86_64+0x79ea9)
    #1 HTTPServerHandler.serialize(_:for:ctx:) HTTPServer.swift:226 (HTTP:x86_64+0x6fda9)
    #2 closure #1 in HTTPServerHandler.respond(to:body:ctx:) HTTPServer.swift:213 (HTTP:x86_64+0x6f964)
...
```
[Full trace in Pastebin](https://pastebin.com/ktyTWQEM)

**Comment by MrMage** [comment]
Thank you for the thread sanitizer info. It indeed looks like would need to be thread-local or synchronized-access.

**Comment by grundoon** [comment]
I believe we have a fix for this, but are waiting on another test case to confirm or utterly dash our hopes for it before opening a PR in vapor/http (bottom line: `gmtime()` is not thread safe, use `gmtime_r()`)

If you'd like to give it a shot in the meantime @anghergh, replace the entire block at https://github.com/vapor/http/blob/master/Sources/HTTP/Utilities/RFC1123.swift#L58-L65 with

   ```swift
    var dateComponents = cachedTimeComponents?.components ?? tm.init()
    if cachedTimeComponents?.key != key {
        gmtime_r(&date, &dateComponents)
        cachedTimeComponents = (key: key, components: dateComponents)
    }

**Comment by MrMage** [comment]
As far as I can tell, `gmtime` is only called once per day. I think the actual fix would be to synchronize access to `cachedTimestamp`, e.g. using a lock. While at it, I would also synchronize access to `cachedTimeComponents`, to avoid a race condition where one thread updates it while another thread reads it. That could cause the second thread to read a stale first half and a fresh second half. Also, we could move `cachedTimeComponents` into `RFC1123DateCache` alongside `cachedTimestamp` and the lock.

**Comment by grundoon** [comment]
Yes, `gmtime` is only called once per day. However, `RFC1123DateCache` is called once per connection: https://github.com/vapor/http/blob/master/Sources/HTTP/Responder/HTTPServer.swift#L130, which is why we're only seeing the race under very heavy load... tons of responses. The `gmtime_r` re-entrant variation *should?* fix that, for now.

I believe the grand plan is to move it off the connection and onto the `EventLoop` itself, based on a suggestion from `@jw`. https://github.com/vapor/http/issues/290

**Comment by MrMage** [comment]
But if `gmtime` is only called once a day, why would moving to a reentrant version fix crashes that occur more often than once per day?

Note: I do agree that moving to `gmtime_r` (and later move this to the event loop) is the right thing to do, I just don't think it will fix all crashes that can happen here.

**Comment by grundoon** [comment]
`gmtime` is only called once per day *per connection* – it never has a chance to live anywhere near that long, only the life of the connection. When the server is essentially being DDoSed by (legitimate) calls and additional threads are firing up, many `RFC1123DateCache`s are being simultaneously created.

**Comment by MrMage** [comment]
Ah, thanks for pointing that out! Sorry, I had thought `RFC1123DateCache` was a singleton, like `cachedTimeComponents`. Please disregard what I said, your fix should indeed work. Sorry again!

**Comment by grundoon** [comment]
No apologies! The `ThreadSpecificVariable` pattern is freaking genius, and will likely solve any remaining vulnerabilities.

**Comment by anghergh** [comment]
@grundoon Thank you for your help! Unfortunately it still crashes in the same file, albeit in the `willCreate` case instead of the `didCreate`. 🤔

<img width="961" alt="screen shot 2018-05-26 at 23 42 56" src="https://user-images.githubusercontent.com/18578250/40580076-fe1d4c4e-613f-11e8-993a-c3aa0094e37c.png">

Except this time I can't get TSan or ASan to trip. 😞 I'll keep trying.

**Comment by grundoon** [comment]
As has been pointed out by another very bright fruit-associated person, there is still a read-write race condition here (as with the original version) and it needs to be wrapped in a sync dispatch. The need of which still irks me since we're trying to eke out performance, but might at least offer @MrMage a certain sense of vindication. :)

**Comment by anghergh** [comment]
Ah I see, I must have missed it. Sorry about that 😅.

**Comment by grundoon** [comment]
No, my error – and it never even crossed my mind that conditional binding is not thread-safe. So either that, or the adapted original

    let dateComponents: tm
    if let cached = cachedTimeComponents, cached.key == key {
        dateComponents = cached.components
    } else {
        var tc = tm.init()
        gmtime_r(&date, &tc)
        dateComponents = tc
        cachedTimeComponents = (key: key, components: tc)
    }

probably just needs to be wrapped as synchronous.

**Comment by MrMage** [comment]
How about just moving `cachedTimeComponents` into the class, so that it is only cached per thread as well, like the final string? (Let me know if I'm missing something again.)

**Comment by anghergh** [comment]
I wrapped it in a sync job, but it still crashed. Then I wrapped the entire function in a global semaphore, and... it still crashed. 

Then I decided to try this
```
    mutating func currentTimestamp() -> String {
        // get the current time

        return "1234"
    }
```
But it doesn't go away. 😕 Either I'm doing something wrong or that's not causing it.
<img width="964" alt="screen shot 2018-05-27 at 01 55 32" src="https://user-images.githubusercontent.com/18578250/40580854-17007504-6151-11e8-9146-bf50235cffa7.png">
<img width="314" alt="screen shot 2018-05-27 at 01 57 17" src="https://user-images.githubusercontent.com/18578250/40580863-6a26c4b8-6151-11e8-800d-0ee3679c403b.png">
Could it have something to do with the ID setter?

**Comment by anghergh** [comment]
Got some TSan traces! 😄
```
==================
WARNING: ThreadSanitizer: data race (pid=38936)
  Write of size 1 at 0x7b20000440f0 by thread T7:
    #0 Request.databaseConnection<A>(to:) Request.swift:172 (Vapor:x86_64+0x7c89d)
    #1 protocol witness for DatabaseConnectable.databaseConnection<A>(to:) in conformance Request Request.swift (Vapor:x86_64+0x7d7cf)
...
```
https://pastebin.com/szTRH95q
<img width="964" alt="screen shot 2018-05-27 at 07 02 19" src="https://user-images.githubusercontent.com/18578250/40582347-186d1ef8-617c-11e8-9b5e-4689a83bcc30.png">

```
==================
WARNING: ThreadSanitizer: Swift access race (pid=38936)
  Read of size 8 at 0x7b080000c2d0 by thread T1:
    #0 Container.requestCachedConnection<A>(to:) Container+CachedConnection.swift:21 (DatabaseKit:x86_64+0x26b8)
    #1 Request.databaseConnection<A>(to:) Request.swift:173 (Vapor:x86_64+0x7c90b)
    #2 protocol witness for DatabaseConnectable.databaseConnection<A>(to:) in conformance Request Request.swift (Vapor:x86_64+0x7d7cf)
...
```
https://pastebin.com/hazm6bnn
<img width="968" alt="screen shot 2018-05-27 at 07 03 14" src="https://user-images.githubusercontent.com/18578250/40582354-2dad99aa-617c-11e8-8e40-1fb6a0a1db65.png">

```
==================
WARNING: ThreadSanitizer: data race (pid=38936)
  Read of size 1 at 0x7b080000c138 by thread T9:
    #0 closure #1 in DatabaseConnectionPool.requestConnection() DatabaseConnectionPool.swift:76 (DatabaseKit:x86_64+0xc012)
    #1 partial apply for closure #1 in DatabaseConnectionPool.requestConnection() DatabaseConnectionPool.swift (DatabaseKit:x86_64+0xc940)
...
```
https://pastebin.com/pdZnd2R9 
<img width="963" alt="screen shot 2018-05-27 at 07 02 34" src="https://user-images.githubusercontent.com/18578250/40582349-2115d82e-617c-11e8-904f-d1dc65518923.png">

**Comment by MrMage** [comment]
Wrong button, sorry.

**Comment by anghergh** [comment]
@MrMage 
> Do you fork out into a different thread in your request handler?

Indeed, I was doing that with a dispatch_async/wait combo, to sync the save of a model with its children array in one request. I refactored the code to only use futures, however it didn't solve the problem. Then I removed the whole children saving code, ending up with this:
(mind the variable names, they made sense in the previous flow)
```
    func create(_ req: Request) throws -> Future<Response> {
        let poll = req.content.get(Poll.self, at: "poll")
        return poll.flatMap { savedPoll -> Future<Response> in
            try savedPoll.validate()
            return savedPoll.save(on: req).map(to: Response.self) { savedPollOptions in
                return req.redirect(to: "/?createPollSuccess=true")
            }
        }
    }
```
Finally TSan tripped with the above mentioned read/write race condition.
<img width="962" alt="screen shot 2018-05-28 at 01 55 45" src="https://user-images.githubusercontent.com/18578250/40591527-833a3e2e-621b-11e8-962d-c3d0201d9996.png">

Being inspired by
> How about just moving `cachedTimeComponents` into the class, so that it is only cached per thread as well, like the final string? (Let me know if I'm missing something again.)

I did that, and TSan doesn't trip anymore. I do get `[ ERROR ] MySQLError.server (1461): Can't create more than max_prepared_stmt_count statements (current value: 16382) (MySQLConnection.swift:71)` after 20-30 seconds, but I saw an existing issue about that, so that's all good.

However, turning TSan off yields the good old:
<img width="960" alt="screen shot 2018-05-28 at 02 08 40" src="https://user-images.githubusercontent.com/18578250/40591561-1230cd28-621c-11e8-9825-1265d0a15888.png">
I ran out of things to remove/refactor at this point 😂
